### PR TITLE
add finality depth contraint in block subscriber

### PR DESCRIPTION
**Pull request by FelixFan1992**

### CODE REVIEW

File: core/services/ocr2/plugins/ocr2keeper/evm21/registry.go [hunk 1]
```diff
@@ -34,7 +34,7 @@ import (
 
 const (
 	// defaultAllowListExpiration decides how long an upkeep's allow list info will be valid for.
-	defaultAllowListExpiration = 20 * time.Minute
+	defaultAllowListExpiration = 5 * time.Minute
```
Comment (infiloop2) [inline core/services/ocr2/plugins/ocr2keeper/evm21/registry.go:37]: can we separate this out as its unrelated to PR

Comment (FelixFan1992) [reply core/services/ocr2/plugins/ocr2keeper/evm21/registry.go:37]: done

File: core/services/ocr2/plugins/ocr2keeper/evm21/streams_lookup.go [hunk 2]
```diff
@@ -410,8 +410,6 @@ func (r *EvmRegistry) singleFeedRequest(ctx context.Context, ch chan<- MercuryDa
 				return fmt.Errorf("at block %s upkeep %s received status code %d for feed %s", sl.time.String(), sl.upkeepId.String(), resp.StatusCode, sl.feeds[index])
 			}
 
-			lggr.Debugf("at block %s upkeep %s received status code %d from mercury v0.2 with BODY=%s", sl.time.String(), sl.upkeepId.String(), resp.StatusCode, hexutil.Encode(body))
```
[the remark this answers is not in the document]
Comment (FelixFan1992) [reply core/services/ocr2/plugins/ocr2keeper/evm21/streams_lookup.go:413]: done

File: core/services/ocr2/plugins/ocr2keeper/evm21/registry_check_pipeline.go [hunk 3]
```diff
@@ -103,6 +103,11 @@ func (r *EvmRegistry) verifyCheckBlock(_ context.Context, checkBlock, upkeepId *
 	var ok bool
 	// verify check block number and hash are valid
 	h, ok = r.bs.queryBlocksMap(checkBlock.Int64())
+	// if this block number/hash combo exists in block subscriber, this check block and hash still exist on chain and are valid
+	if ok && h == checkHash.Hex() {
+		r.lggr.Debugf("check block hash %s exists on chain at block number %d for upkeepId %s", checkHash.Hex(), checkBlock, upkeepId)
+		return encoding.NoPipelineError, false
+	}
 	if !ok {
```
Comment (infiloop2) [inline core/services/ocr2/plugins/ocr2keeper/evm21/registry_check_pipeline.go:111]: we'll need to remove this if condition, we want to do also check in the case hash existed but didn't match

File: core/services/ocr2/plugins/ocr2keeper/evm21/registry_check_pipeline.go [hunk 4]
```diff
@@ -103,6 +103,11 @@ func (r *EvmRegistry) verifyCheckBlock(_ context.Context, checkBlock, upkeepId *
 	var ok bool
 	// verify check block number and hash are valid
 	h, ok = r.bs.queryBlocksMap(checkBlock.Int64())
+	// if this block number/hash combo exists in block subscriber, this check block and hash still exist on chain and are valid
```
Comment (infiloop2) [inline core/services/ocr2/plugins/ocr2keeper/evm21/registry_check_pipeline.go:106]: can you expand a bit here along the lines

"the block hash in block subscriber might be slightly outdated, if it doesn't match then we fetch the latest from RPC"

File: core/services/ocr2/plugins/ocr2keeper/evm21/registry_check_pipeline.go [hunk 5]
```diff
@@ -103,6 +103,11 @@ func (r *EvmRegistry) verifyCheckBlock(_ context.Context, checkBlock, upkeepId *
...
+	if ok && h == checkHash.Hex() {
+		r.lggr.Debugf("check block hash %s exists on chain at block number %d for upkeepId %s", checkHash.Hex(), checkBlock, upkeepId)
+		return encoding.NoPipelineError, false
+	}
 	if !ok {
 		r.lggr.Warnf("check block %s does not exist in block subscriber for upkeepId %s, querying eth client", checkBlock, upkeepId)
```
Comment (infiloop2) [inline core/services/ocr2/plugins/ocr2keeper/evm21/registry_check_pipeline.go:112]: will need to reword this log too

File: core/services/ocr2/plugins/ocr2keeper/evm21/registry_check_pipeline.go [hunk 6]
```diff
@@ -126,10 +131,13 @@ func (r *EvmRegistry) verifyLogExists(upkeepId *big.Int, p ocr2keepers.UpkeepPay
 	// if log block number is populated, check log block number and block hash
 	if logBlockNumber != 0 {
 		h, ok := r.bs.queryBlocksMap(logBlockNumber)
+		// if this block number/hash combo exists in block subscriber, this block and tx still exists on chain and is valid
```
Comment (infiloop2) [inline core/services/ocr2/plugins/ocr2keeper/evm21/registry_check_pipeline.go:134]: similarly can we expand on the comment on why we ignore block subscriber's hash when it existed

**Review by infiloop2 — CHANGES_REQUESTED** [review]
requesting changes for if/else logic in verifyCheckBlock

File: core/services/ocr2/plugins/ocr2keeper/evm21/registry_check_pipeline.go (hunk 6 above)
Comment (FelixFan1992) [reply core/services/ocr2/plugins/ocr2keeper/evm21/registry_check_pipeline.go:134]: done

File: core/services/ocr2/plugins/ocr2keeper/evm21/registry_check_pipeline.go (hunk 5 above)
Comment (FelixFan1992) [reply core/services/ocr2/plugins/ocr2keeper/evm21/registry_check_pipeline.go:112]: done

**Review by infiloop2 — APPROVED** [review]

**Pull request merged**